bson/primitive/decimal.go

```diff
@@ -139,169 +220,117 @@ func dErr(s string) (Decimal128, error) {
-			}
-			mul *= 10
-			add *= 10
-			add += uint32(c - '0')
-			continue
-		}
-		if c == '.' {
-			i++
-			if dot >= 0 || i == 1 && len(s) == 1 {
-				return dErr(orig)
-			}
-			if i == len(s) {
-				break
-			}
-			if s[i] < '0' || s[i] > '9' || e > 0 {
-				return dErr(orig)
-			}
-			dot = i
-			continue
+	ip := matches[1]
+	dp := matches[2]
+	ep := matches[3]
+
+	var err error
+	e := 0
+	if ep != "" {
+		e, err = strconv.Atoi(ep)
+		if err != nil {
+			return dErr(s)
 		}
-		break
 	}
-	if i == 0 {
-		return dErr(orig)
+	if dp != "" {
+		e -= len(dp)
 	}
-	if mul > 1 {
-		h, l, ovr = muladd(h, l, mul, add)
-		if ovr > 0 || h&((1<<15-1)<<49) > 0 {
-			return dErr(orig)
-		}
+
+	if len(strings.Trim(ip+dp, "-0")) > 35 {
+		return dErr(s)
 	}
-	if dot >= 0 {
-		e += dot - i
+
+	bi, ok := new(big.Int).SetString(ip+dp, 10)
+	if !ok {
+		return dErr(s)
 	}
-	if i+1 < len(s) && (s[i] == 'E' || s[i] == 'e') {
-		i++
-		eneg := s[i] == '-'
-		if eneg || s[i] == '+' {
-			i++
-			if i == len(s) {
-				return dErr(orig)
-			}
-		}
-		n := 0
-		for i < len(s) && n < 1e4 {
-			c := s[i]
-			i++
-			if c < '0' || c > '9' {
-				return dErr(orig)
-			}
-			n *= 10
-			n += int(c - '0')
-		}
-		if eneg {
-			n = -n
+
+	d, ok := ParseDecimal128FromBigInt(bi, e)
+	if !ok {
+		return dErr(s)
+	}
+	return d, nil
+}
+
+var (
+	ten  = big.NewInt(10)
+	zero = new(big.Int)
+	maxS = new(big.Int).SetBytes([]byte{0x1, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}) // 113 bits
+)
+
+// ParseDecimal128FromBigInt attempts to parse the given significand and exponent into a valid Decimal128 value.
+func ParseDecimal128FromBigInt(bi *big.Int, exp int) (Decimal128, bool) {
+	//copy
+	if bi == nil {
+		bi = new(big.Int)
+	} else {
+		bi = new(big.Int).Set(bi)
+	}
+
+	q := new(big.Int)
+	r := new(big.Int)
+
+	for bi.CmpAbs(maxS) == 1 {
+		bi, _ = q.QuoRem(bi, ten, r)
+		if r.Cmp(zero) != 0 {
+			return Decimal128{}, false
 		}
-		e += n
-		for e < -6176 {
-			// Subnormal.
-			var div uint32 = 1
-			for div < 1e9 && e < -6176 {
-				div *= 10
-				e++
-			}
-			var rem uint32
-			h, l, rem = divmod(h, l, div)
-			if rem > 0 {
-				return dErr(orig)
-			}
+		exp++
+		if exp > MaxDecimal128Exp {
+			return Decimal128{}, false
 		}
-		for e > 6111 {
-			// Clamped.
-			var mul uint32 = 1
-			for mul < 1e9 && e > 6111 {
-				mul *= 10
-				e--
-			}
-			h, l, ovr = muladd(h, l, mul, 0)
-			if ovr > 0 || h&((1<<15-1)<<49) > 0 {
-				return dErr(orig)
-			}
+	}
+
+	for exp < MinDecimal128Exp {
+		// Subnormal.
+		bi, _ = q.QuoRem(bi, ten, r)
+		if r.Cmp(zero) != 0 {
+			return Decimal128{}, false
 		}
-		if e < -6176 || e > 6111 {
-			return dErr(orig)
+		exp++
+	}
+	for exp > MaxDecimal128Exp {
+		// Clamped.
+		bi.Mul(bi, ten)
+		if bi.CmpAbs(maxS) == 1 {
+			return Decimal128{}, false
 		}
+		exp--
 	}
 
-	if i < len(s) {
-		return dErr(orig)
+	b := bi.Bytes()
+	var h, l uint64
+	for i := 0; i < len(b); i++ {
+		if i < len(b)-8 {
```

Comment:
This can be rewritten with a continue in the if statement and then the else statement can be removed:
```go
if i < len(b)-8 {
    h = h<<8 | uint64(b[i])
    continue
}
l = l<<8 | uint64(b[i])
```